File: src-ts/services/detector.ts
/**
 * Content Detection Service
 *
 * Detects missing content and content below quality cutoff across all
 * configured Radarr and Sonarr servers.
 */

import type { ServerConfig, DetectionResult } from "../types";
import { RadarrClient, SonarrClient, createClient } from "../lib/api-client";
import { getDatabase } from "../storage/database";

/** Aggregated detection results across all servers */
export interface AggregatedResults {
  results: DetectionResult[];
  totalMissing: number;
  totalCutoff: number;
  successCount: number;
  failureCount: number;
}

/**
 * Detect missing and cutoff content on a single server (internal)
 */
async function detectOnServer(server: ServerConfig): Promise<DetectionResult> {
  const client = createClient(server.url, server.apiKey, server.type);

  const result: DetectionResult = {
    serverId: server.id,
    serverName: server.name,
    serverType: server.type,
    missingCount: 0,
    cutoffCount: 0,
    missingItems: [],
    cutoffItems: [],
  };

  // Get missing items
  const missingResult =
    server.type === "radarr"
      ? await (client as RadarrClient).getAllMissing()
      : await (client as SonarrClient).getAllMissing();

  if (missingResult.success) {
    result.missingItems = missingResult.data;
    result.missingCount = missingResult.data.length;
  } else {
    result.error = `Missing detection failed: ${missingResult.error}`;
    return result;
  }

  // Get cutoff unmet items
  const cutoffResult =
    server.type === "radarr"
      ? await (client as RadarrClient).getAllCutoffUnmet()
      : await (client as SonarrClient).getAllCutoffUnmet();

  if (cutoffResult.success) {
    result.cutoffItems = cutoffResult.data;
    result.cutoffCount = cutoffResult.data.length;
  } else {
    result.error = `Cutoff detection failed: ${cutoffResult.error}`;
    return result;
  }

  return result;
}

/**
 * Run detection on all configured servers
 */
export async function detectAll(): Promise<AggregatedResults> {
  const db = getDatabase();
  const servers = await db.getAllServers();

  const results: DetectionResult[] = [];
  let totalMissing = 0;
  let totalCutoff = 0;
  let successCount = 0;
  let failureCount = 0;

  // Run detection on all servers concurrently
  const detectionPromises = servers.map((server) => detectOnServer(server));
  const detectionResults = await Promise.all(detectionPromises);

  for (const result of detectionResults) {
    results.push(result);

    if (result.error) {
      failureCount++;
    } else {
      successCount++;
      totalMissing += result.missingCount;
      totalCutoff += result.cutoffCount;
    }
  }

  return {
    results,
    totalMissing,
    totalCutoff,
    successCount,
    failureCount,
  };
}

/**
 * Run detection on servers of a specific type
 */
export async function detectByType(
  type: "radarr" | "sonarr"
): Promise<AggregatedResults> {
  const db = getDatabase();
  const servers = await db.getServersByType(type);

  const results: DetectionResult[] = [];
  let totalMissing = 0;
  let totalCutoff = 0;
  let successCount = 0;
  let failureCount = 0;

  const detectionPromises = servers.map((server) => detectOnServer(server));
  const detectionResults = await Promise.all(detectionPromises);

  for (const result of detectionResults) {
    results.push(result);

    if (result.error) {
      failureCount++;
    } else {
      successCount++;
      totalMissing += result.missingCount;
      totalCutoff += result.cutoffCount;
    }
  }

  return {
    results,
    totalMissing,
    totalCutoff,
    successCount,
    failureCount,
  };
}

/**
 * Run detection on a single server by ID or name
 */
export async function detectSingleServer(
  idOrName: string
): Promise<DetectionResult | null> {
  const db = getDatabase();

  // Try by ID first
  let server = await db.getServer(idOrName);

  // Then try by name
  if (!server) {
    server = await db.getServerByName(idOrName);
  }

  if (!server) {
    return null;
  }

  return detectOnServer(server);
}

File: src/services/detector.go
package services

import (
	"context"
	"fmt"
	"sync"

	"github.com/user/janitarr/src/api"
	"github.com/user/janitarr/src/database"
)

// DetectorAPIClient is the interface for API clients used by the Detector.
type DetectorAPIClient interface {
	TestConnection(ctx context.Context) (*api.SystemStatus, error)
	GetAllMissing(ctx context.Context) ([]api.MediaItem, error)
	GetAllCutoffUnmet(ctx context.Context) ([]api.MediaItem, error)
	TriggerSearch(ctx context.Context, ids []int) error
}

// DetectorAPIClientFactory creates API clients for detection.
type DetectorAPIClientFactory func(url, apiKey, serverType string) DetectorAPIClient

// defaultDetectorAPIClientFactory creates real API clients.
func defaultDetectorAPIClientFactory(url, apiKey, serverType string) DetectorAPIClient {
	if serverType == "sonarr" {
		return api.NewSonarrClient(url, apiKey)
	}
	return api.NewRadarrClient(url, apiKey)
}

// Detector detects missing content and content below quality cutoff across all servers.
type Detector struct {
	db         *database.DB
	apiFactory DetectorAPIClientFactory
}

// NewDetector creates a new Detector with the given database.
func NewDetector(db *database.DB) *Detector {
	return &Detector{
		db:         db,
		apiFactory: defaultDetectorAPIClientFactory,
	}
}

// NewDetectorWithFactory creates a new Detector with a custom API factory.
// Useful for testing.
func NewDetectorWithFactory(db *database.DB, factory DetectorAPIClientFactory) *Detector {
	return &Detector{
		db:         db,
		apiFactory: factory,
	}
}

// DetectAll runs detection on all enabled servers concurrently.
func (d *Detector) DetectAll(ctx context.Context) (*DetectionResults, error) {
	servers, err := d.db.GetAllServers()
	if err != nil {
		return nil, fmt.Errorf("getting servers: %w", err)
	}

	// Filter to only enabled servers
	var enabledServers []database.Server
	for _, s := range servers {
		if s.Enabled {
			enabledServers = append(enabledServers, s)
		}
	}

	// Return empty results if no servers
	if len(enabledServers) == 0 {
		return &DetectionResults{
			Results:      []DetectionResult{},
			TotalMissing: 0,
			TotalCutoff:  0,
			SuccessCount: 0,
			FailureCount: 0,
		}, nil
	}

	// Run detection on all servers concurrently
	var wg sync.WaitGroup
	resultCh := make(chan DetectionResult, len(enabledServers))

	for _, server := range enabledServers {
		wg.Add(1)
		go func(s database.Server) {
			defer wg.Done()
			result := d.detectServer(ctx, &s)
			resultCh <- result
		}(server)
	}

	// Wait for all detections to complete
	wg.Wait()
	close(resultCh)

	// Aggregate results
	results := &DetectionResults{
		Results: make([]DetectionResult, 0, len(enabledServers)),
	}

	for result := range resultCh {
		results.Results = append(results.Results, result)

		if result.Error != "" {
			results.FailureCount++
		} else {
			results.SuccessCount++
			results.TotalMissing += len(result.Missing)
			results.TotalCutoff += len(result.Cutoff)
		}
	}

	return results, nil
}

// detectServer runs detection on a single server.
func (d *Detector) detectServer(ctx context.Context, server *database.Server) DetectionResult {
	result := DetectionResult{
		ServerID:     server.ID,
		ServerName:   server.Name,
		ServerType:   string(server.Type),
		Missing:      []int{},
		Cutoff:       []int{},
		MissingItems: make(map[int]api.MediaItem),
		CutoffItems:  make(map[int]api.MediaItem),
	}

	client := d.apiFactory(server.URL, server.APIKey, string(server.Type))

	// Get missing items
	missingItems, err := client.GetAllMissing(ctx)
	if err != nil {
		result.Error = fmt.Sprintf("missing detection failed: %v", err)
		return result
	}

	for _, item := range missingItems {
		result.Missing = append(result.Missing, item.ID)
		result.MissingItems[item.ID] = item
	}

	// Get cutoff unmet items
	cutoffItems, err := client.GetAllCutoffUnmet(ctx)
	if err != nil {
		result.Error = fmt.Sprintf("cutoff detection failed: %v", err)
		return result
	}

	for _, item := range cutoffItems {
		result.Cutoff = append(result.Cutoff, item.ID)
		result.CutoffItems[item.ID] = item
	}

	return result
}

// DetectServer runs detection on a single server by ID.
func (d *Detector) DetectServer(ctx context.Context, serverID string) (*DetectionResult, error) {
	server, err := d.db.GetServer(serverID)
	if err != nil {
		return nil, err
	}
	if server == nil {
		return nil, fmt.Errorf("server not found: %s", serverID)
	}

	result := d.detectServer(ctx, server)
	return &result, nil
}

// DetectByType runs detection on all enabled servers of a specific type.
func (d *Detector) DetectByType(ctx context.Context, serverType database.ServerType) (*DetectionResults, error) {
	servers, err := d.db.GetServersByType(serverType)
	if err != nil {
		return nil, fmt.Errorf("getting servers by type: %w", err)
	}

	// Filter to only enabled servers
	var enabledServers []database.Server
	for _, s := range servers {
		if s.Enabled {
			enabledServers = append(enabledServers, s)
		}
	}

	// Return empty results if no servers
	if len(enabledServers) == 0 {
		return &DetectionResults{
			Results:      []DetectionResult{},
			TotalMissing: 0,
			TotalCutoff:  0,
			SuccessCount: 0,
			FailureCount: 0,
		}, nil
	}

	// Run detection concurrently
	var wg sync.WaitGroup
	resultCh := make(chan DetectionResult, len(enabledServers))

	for _, server := range enabledServers {
		wg.Add(1)
		go func(s database.Server) {
			defer wg.Done()
			result := d.detectServer(ctx, &s)
			resultCh <- result
		}(server)
	}

	wg.Wait()
	close(resultCh)

	// Aggregate results
	results := &DetectionResults{
		Results: make([]DetectionResult, 0, len(enabledServers)),
	}

	for result := range resultCh {
		results.Results = append(results.Results, result)

		if result.Error != "" {
			results.FailureCount++
		} else {
			results.SuccessCount++
			results.TotalMissing += len(result.Missing)
			results.TotalCutoff += len(result.Cutoff)
		}
	}

	return results, nil
}
